### ISSUE TITLE
Class should have a primary constructor

### ISSUE THREAD

**Issue by pedromassango** [dart-lang/language#138]
Why not to have a primary constructor?
**This way:**
```
class Person(String firstName, Int age) { 

    // primary constructor initialization (optional)
    init{

    }
}
```

**Is better than this:**

```
class Person { 

String firstName;
int age;

    Person(this.firstName, this.age){
        // constructor initialization
    }
}
```

As you can see in Dart we need to create the property inside de class and then call in our constructor, would be good to create the property inside the primary constructor as we touch the property only once and it will be available in whole class.

**Comment by munificent** [comment]
I don't think the fields should be right after the class name. In many cases, there is too much other syntax already right there. The class name may be long, may have type parameters (which may have bounds). There can be an `extends` clause, `implements` clause (with possibly many interfaces), and an `on` clause (again, possibly with many mixins).

Stuffing fields in there too is a *lot* of syntax to get through before you hit the `{`. This makes it harder to find the field declarations, and disincentivizes writing doc comments for them. The latter is, I think, particularly problematic.

**Comment by pedromassango** [comment]
Have you used Kotlin?

**Comment by munificent** [comment]
A little. I've already read a bunch about a number of other languages that do primary constructors, as well as designing a hobby language or two that work that way. I think it's nice for small classes, but it scales up poorly.

**Comment by yjbanov** [comment]
I agree that this syntax doesn't scale well to arbitrary classes. However, it would be very useful for things like complex return types. There's an [alternative proposal](https://github.com/dart-lang/language/issues/68) for this, but I think if we had a low-boilerplate solution for declaring "value"/"data" classes we wouldn't need multiple return types or primary constructors.

If we get [immutability](https://github.com/dart-lang/language/issues/125) with a [nice syntax](https://github.com/dart-lang/language/issues/125) and [good semantics](https://github.com/dart-lang/language/issues/125#issuecomment-444697727) I think we will solve this issue as well.

**Comment by mraleph** [comment]
> I think it's nice for small classes, but it scales up poorly.

Conversely things that scale to large classes are usually too unwieldy for small classes. 

I think there is a value giving developers an ability to declare small classes without pages of ceremonious code.

**Comment by pedromassango** [comment]
@munificent Think of me, Kotlin's way and the current Dart way, which one do you think is best for classes with more than 10 parameters?

The Dart way the developer has to write so much unnecessarily. Take a look at Kotlin's way and I think you'll give me some reason.

**Comment by yjbanov** [comment]
@mraleph 

> Conversely things that scale to large classes are usually too unwieldy for small classes.

You are as good at minimizing the number of words to express something that takes 5x the number of words for me, as you are at minimizing the number of CPU instructions to express programs :smile:

**Comment by munificent** [comment]
> I think there is a value giving developers an ability to declare small classes without pages of ceremonious code.

I agree, but I like it best when those features scale smoothly to larger use cases. Many classes that start small end up big, and it's no fun if you have to eventually discard the primary constructor syntax and rewrite the entire class using the other syntax. Incremental changes to a class should require incremental changes to the program text.

For example (I haven't put more than five minutes of thought into this), we could say that if you define a class like:

```dart
class Person {
  String firstName;
  int age;
}
```

Then it implicitly gets a constructor that takes those fields as named (or maybe positional?) parameters. This gives you an easy way to get a small class up and running. If you later decide to do some validation, you can add a constructor, but you never have to rearrange the code to move the fields out of a primary constructor and into the body.

**Comment by patelkrunal78** [comment]
Similar Issue To [Dart Multiple Constructors](https://icetutor.com/question/dart-multiple-constructors)

**Comment by derolf** [comment]
Please either adopt the syntax of:
- Scala case classes
- Kotlin data classes
- Typescript constructor instance parameters

No need to invent something completely new!

**Comment by kevmoo** [comment]
See https://github.com/dart-lang/language/issues/698 – I think gets you where you want to go!

**Comment by leafpetersen** [comment]
We're tracking this request [here](https://github.com/dart-lang/language/issues/2364), closing in favor of that issue.
